Sources/Datadog/RUM/RUMVitals/VitalCPUReader.swift

```diff
@@ -0,0 +1,82 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+import UIKit.UIApplication
+
+/// A class reading the CPU ticks (_1 second = 600 ticks_) since the start of the process.
```

Comment:
Where does _600 ticks_ come from? Won't it differ per CPU?

Reply:
TL;DR: i looked for `counter`s to get executed instructions count or anything like that. however, kernel returns us `timer`s even at the lowest point and tells us _how long_ CPU spent in user/sys/idle mode.

i couldn't debug the kernel code but as far as i could follow [the code](https://github.com/opensource-apple/xnu/blob/master/osfmk/kern/host.c#L436):
here is the macro that computes `cpu_ticks`
```
#define GET_TICKS_VALUE_FROM_TIMER(processor, state, timer) {
  GET_TICKS_VALUE(state, timer_grab(&PROCESSOR_DATA(processor, timer)));
}
```
`GET_TICKS_VALUE` doesn't do much, divides `ticks` _probably_ by 10^7 (coming from [here](https://github.com/opensource-apple/xnu/blob/master/osfmk/kern/clock.c#L367)) and stores it in `cpu_load_info`
```
#define GET_TICKS_VALUE(state, ticks) {
  cpu_load_info->cpu_ticks[(state)] += (uint32_t)(ticks / hz_tick_interval);
}
```
[`timer_grab`](https://github.com/opensource-apple/xnu/blob/master/osfmk/kern/timer.h#L148) takes a `timer_t` and returns its `all_bits`, that should mean something like timer's timestamp i guess
```
static inline uint64_t timer_grab(timer_t timer) { return timer->all_bits; }
```
[`PROCESSOR_DATA`](https://github.com/opensource-apple/xnu/blob/master/osfmk/kern/processor_data.h#L116) macro takes a `processor` and returns the `timer_t` corresponding to given `member`
`member` can be `user_state`/`system_state`/`idle_state`
```
#define PROCESSOR_DATA(processor, member) { (processor)->processor_data.member }
```
and processor_data struct has a bunch of timers
```
struct processor_data {
  timer_data_t idle_state;
  ...
}
```
and that's the lowest i could get.

Reply:
I see 👌 , this is pretty complex to me as I don't know into these details. I made few more tests on my iPhone X, receiving some failures - explained in comment below.